tensorflow_addons/optimizers/discriminative_layer_training_test.py

```diff
@@ -0,0 +1,271 @@
+        d_opt = DiscriminativeWrapper(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        self._assert_training_losses_are_close(model, model_lr)
+
+    def _test_equal_0_layer_lr_to_trainable_false(self, model_fn, loss, opt):
+        """confirm 0 lr_mult for the model is the same as model not trainable"""
+        learning_rate = 0.01
+        model = model_fn()
+        model.trainable = False
+        model.compile(loss=loss, optimizer=opt(learning_rate))
+
+        model_lr = model_fn()
+        model_lr.lr_mult = 0.0
+        d_opt = DiscriminativeWrapper(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        self._assert_training_losses_are_close(model, model_lr)
+
+    def _test_equal_half_layer_lr_to_half_lr_of_opt(self, model_fn, loss, opt):
+        """confirm 0.5 lr_mult for the model is the same as optim with 0.5 lr"""
+
+        mult = 0.5
+        learning_rate = 0.01
+        model = model_fn()
+        model.compile(loss=loss, optimizer=opt(learning_rate * mult))
+
+        model_lr = model_fn()
+        model_lr.lr_mult = mult
+        d_opt = DiscriminativeWrapper(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        self._assert_training_losses_are_close(model, model_lr)
+
+    def _test_loss_changes_over_time(self, model_fn, loss, opt):
+        """confirm that model trains with lower lr on specific layer"""
+
+        learning_rate = 0.01
+        model_lr = model_fn()
+        model_lr.layers[0].lr_mult = 0.01
+        d_opt = DiscriminativeWrapper(
+            opt, model_lr, verbose=False, learning_rate=learning_rate
+        )
+        model_lr.compile(loss=loss, optimizer=d_opt)
+
+        loss_values = get_losses(get_train_results(model_lr))
+        self.assertLess(loss_values[-1], loss_values[0])
+
+    def _run_tests_in_notebook(self):
+        for name, method in DiscriminativeLearningTest.__dict__.items():
+            if callable(method) and name[:4] == "test":
+                print("running test %s" % name)
+                method(self)
+
+
+def run_distributed(devices):
+    def decorator(f):
+        def decorated(self, *args, **kwargs):
+            logical_devices = devices
+            strategy = tf.distribute.MirroredStrategy(logical_devices)
+            with strategy.scope():
+                f(self, *args, **kwargs)
+
+        return decorated
+
+    return decorator
+
+
+def test_wrap(method, devices, **kwargs):
+    @test_utils.run_in_graph_and_eager_modes
+    def single(self):
+        return method(self, **kwargs)
+
+    @test_utils.run_in_graph_and_eager_modes
+    @run_distributed(devices)
+    def distributed(self):
+        return method(self, **kwargs)
+
+    return single, distributed
+
+
```

Comment:
Please add tests which define learning_rate multipliers and assess that using the Wrapper / as well as Model manager is correctly setting the var attributes. For example set the top layer and bottom layer and verify that only they are set.

Reply:
good point. I'll explicitly test to verify that multipliers are set properly for the variables based on the layer's lr_mult